fableous-be/handlers/websocket.go

```diff
@@ -44,76 +65,136 @@ func (m *module) ConnectHubWS(ctx *gin.Context, classroomID string) (err error)
 	return m.HubCommandWorker(conn, sess)
 }
 
-func (m *module) ConnectControllerWS(ctx *gin.Context, classroomToken, role string) (err error) {
-	var sess *session
-	if sess = m.GetClassroomSession(classroomToken); sess == nil {
-		log.Printf("session not initialised")
+func (m *module) ConnectControllerWS(ctx *gin.Context, classroomToken, role, name string) (err error) {
+	ctx.Request.Header.Del("Sec-Websocket-Extensions")
+	var sess *activeSession
+	if sess = m.GetClassroomActiveSession(classroomToken); sess == nil {
+		log.Println("session not activated")
 		return
 	}
-	if conn := m.GetSessionController(sess, role); conn != nil {
-		log.Printf("role already connected")
+	if conn := m.GetActiveSessionController(sess, role); conn != nil {
+		log.Println("role already connected")
 		return
 	}
-	ctx.Request.Header.Del("Sec-Websocket-Extensions")
-	upgrader := websocket.Upgrader{
-		CheckOrigin: func(r *http.Request) bool {
-			return true
-		},
-	}
 	var conn *websocket.Conn
-	if conn, err = upgrader.Upgrade(ctx.Writer, ctx.Request, nil); err != nil {
+	if conn, err = m.upgrader.Upgrade(ctx.Writer, ctx.Request, nil); err != nil {
 		log.Printf("failed connecting controller websocket. %s\n", err)
 		return
 	}
 	defer conn.Close()
 	sess.mutex.Lock()
 	sess.controllerConn[role] = conn
 	sess.mutex.Unlock()
-	return m.ControllerCommandWorker(conn, sess, role)
+	return m.ControllerCommandWorker(conn, sess, role, name)
 }
 
-func (m *module) HubCommandWorker(conn *websocket.Conn, sess *session) (err error) {
+func (m *module) HubCommandWorker(conn *websocket.Conn, sess *activeSession) (err error) {
+	defer func() {
+		delete(m.sessions.keys, sess.classroomToken)
```

Comment:
wanted to confirm, no locking is needed here when deleting?

Reply:
fixed